### ISSUE TITLE
Distance curve to line

### ISSUE THREAD

**Issue by JoseConseco** [LeanderSilur/Snippets#1]
Hello
I came here from: [stack overflow](https://stackoverflow.com/questions/2742610/closest-point-on-a-cubic-bezier-curve)
Can same technique be applied to curve - line distance? I tried but it seems to give bad results, maybe I'm doing something wrong:
`line_3d(t) = **p** + **r***t`

For 3D space, **P**- point on line, **R**- direction vector of line 3d
Distance squared: 
    `distance'(t)^2 = (bezier_x(t) - line_x(t))^2 + (bezier_y(t) - line_y(t))^2:`
Symbolic solving for extrema of distance'(t):
    `diff((ax*t^3+bx*t^2+cx*t+dx-(px+rx*t))^2,t)`
Gave:
    `6*ax^2*t^5+10*ax*bx*t^4+(-8*ax*rx+8*ax*cx+4*bx^2)*t^3+(-6*bx*rx-6*ax*px+6*ax*dx+6*bx*cx)*t^2+(2*rx^2-4*cx*rx-4*bx*px+4*bx*dx+2*cx^2)*t+(2*px-2*dx)*rx-2*cx*px+2*cx*dx`
Rest is same as in your code, except line for selecting minimal distance:
    `distances = np.sum(np.square(points - (line_dir*potentials[:, None] + line_pt)), axis=1)`

Any ideas why it wont show correct point (it shows point on 3d bezier curve, but not close to line(t).
Sample code:

[spline_line.txt](https://github.com/LeanderSilur/Snippets/files/4777115/spline_line.txt)

Hope it is not trouble.

**Comment by LeanderSilur** [comment]
Interesting, I will take a look at it soon. Line-Bezier seems trivial (I believe your approach should work), but that are a lot of variables. It's wrong differentiation most of the time.

@JoseConseco Could you briefly confirm that the point-bezier also works for 3D space?
And another thing, I didn't check the numbers myself, but wolfram alpha gets a [different result](https://www.wolframalpha.com/input/?i=d%2Fdt+%28ax*t%5E3%2Bbx*t%5E2%2Bcx*t%2Bdx-%28px%2Brx*t%29%29%5E2) than you.

**Comment by JoseConseco** [comment]
Yes, the code for 2d bezier-point distance  works for 3d curve, too:
![3dCurve_point_dist](https://user-images.githubusercontent.com/13521338/84604750-7e924a80-ae98-11ea-8a60-c94c142c6af3.gif)
I used maxima for differentiation, but will see. Maybe I made mistake somewhere
![image](https://user-images.githubusercontent.com/13521338/84604778-ae415280-ae98-11ea-80e0-7d307459a949.png)
I used above code, in the: spline_line.txt, that I have sent you above.

**Comment by JoseConseco** [comment]
I think I figured out why it wont work. lets say we have two lines, and we search closest points:
L1(t) = (0,0,0) + (0,0,1) * t
L2(t) = (1,0,0) + (0,0,-1) * t
They cross at (0,0,0), but not when t = 0. If you evaluate L1(t=0) and L2(t=0), we get two points:
(0,0,0) and (1,0,0) - and distance between those points is not min.
Actual min distance calculated from diff(dist(L1(t), L2(t)) is for t=1/2, but it is not (0,0,0).
![image](https://user-images.githubusercontent.com/13521338/84656021-4fb8ba80-af12-11ea-9e61-594954075bf4.png)

This makes sense, since for t=0, we get distance(L1(0), L2(0)) = 1
And for t=1/2 we get smaller distance (I think it is minimum). 
So yes I'm not sure if tere is way to obtain closest points on curves: maybe using L1(t1), and L2(t2) and solve min dist for (t1,t2) ?

**Comment by LeanderSilur** [comment]
It doesn't work, because your only using one partial derivate, but you have a minimal problem with two variables. I'm working on a line-bezier function distance right now.

There is a good chance, that I don't understand your formulas correctly though.

**Comment by JoseConseco** [comment]
yes, that is why it did not work in Bezier- line - I used one variable t, but it should be t1,t2

**Comment by LeanderSilur** [comment]
You need to take `f(t, r) = |(a t^3 + b t^2 + c t + d - p - rq)|` and find `d/dt = 0 AND d/dr = 0`. However, this includes [substituting](https://www.youtube.com/watch?v=bMlA7WY2LvA) d/dr into d/dt and that looks like a lot of numbers. Will look into it in the future.

**Comment by JoseConseco** [comment]
I did the math, but the problem is when doing 
`diff( dist(t1,t2) , d_t1) `
and 
`diff( dist(t1,t2) , d_t2) `
Results  in  2 polynomial equations and both of them use  t1, t2 variables.
so I cant calculate roots for any of them. It looks like its impossible to solve.

**Comment by LeanderSilur** [comment]
At that point you substitute either (t1 or t2) from one equation to the other.

**Comment by JoseConseco** [comment]
I tried but substituting t2 (solution of d/dt2==0), into first polynomial gives 0:
![image](https://user-images.githubusercontent.com/13521338/84690047-316ab300-af42-11ea-805e-ceefa80f5db9.png)
Which is quite easy to see why in image above. I think found another way to do what I want (project 3d curve to 2d view, and then use 2d curve * point  method).

**Comment by LeanderSilur** [comment]
@JoseConseco Funny, I was just going to suggest this workaround to you.

**Comment by JoseConseco** [comment]
I still wonder why above wont work (the 'deeper' meaning of this) but well, I guess I will never know. Btw. I thanks for help - I used the spline-point in my addon for modeling cloth. Let me know if you want copy (it for making sewing patterns like Marvelous designer).
